source/opt/scalar_replacement_pass.cpp

```diff
@@ -0,0 +1,702 @@
+  switch (type->opcode()) {
+    case SpvOpTypeStruct:
+      type->ForEachInOperand([this, inst, &elem, replacements](uint32_t* id) {
+        CreateVariable(*id, inst, elem++, replacements);
+      });
+      break;
+    case SpvOpTypeArray:
+      for (uint32_t i = 0; i != GetArrayLength(type); ++i) {
+        CreateVariable(type->GetSingleWordInOperand(0u), inst, i, replacements);
+      }
+      break;
+
+    case SpvOpTypeMatrix:
+    case SpvOpTypeVector:
+      for (uint32_t i = 0; i != GetNumElements(type); ++i) {
+        CreateVariable(type->GetSingleWordInOperand(0u), inst, i, replacements);
+      }
+      break;
+
+    default:
+      assert(false && "Unexpected type.");
+      break;
+  }
+
+  TransferAnnotations(inst, replacements);
+}
+
+void ScalarReplacementPass::TransferAnnotations(
+    const ir::Instruction* source,
+    std::vector<ir::Instruction*>* replacements) {
+  // Only transfer invariant and restrict decorations on the variable. There are
+  // no type or member decorations that are necessary to transfer.
+  for (auto inst :
+       get_decoration_mgr()->GetDecorationsFor(source->result_id(), false)) {
+    assert(inst->opcode() == SpvOpDecorate);
+    uint32_t decoration = inst->GetSingleWordInOperand(1u);
+    if (decoration == SpvDecorationInvariant ||
+        decoration == SpvDecorationRestrict) {
+      for (auto var : *replacements) {
+        std::unique_ptr<ir::Instruction> annotation(new ir::Instruction(
+            context(), SpvOpDecorate, 0, 0,
+            std::initializer_list<ir::Operand>{
+                {SPV_OPERAND_TYPE_ID, {var->result_id()}},
+                {SPV_OPERAND_TYPE_DECORATION, {decoration}}}));
+        for (uint32_t i = 2; i < inst->NumInOperands(); ++i) {
+          ir::Operand copy(inst->GetInOperand(i));
+          annotation->AddOperand(std::move(copy));
+        }
+        context()->AddAnnotationInst(std::move(annotation));
+        get_def_use_mgr()->AnalyzeInstUse(&*--context()->annotation_end());
+      }
+    }
+  }
+}
+
+void ScalarReplacementPass::CreateVariable(
+    uint32_t typeId, ir::Instruction* varInst, uint32_t index,
+    std::vector<ir::Instruction*>* replacements) {
+  uint32_t ptrId = GetOrCreatePointerType(typeId);
+  uint32_t id = TakeNextId();
+  std::unique_ptr<ir::Instruction> variable(new ir::Instruction(
+      context(), SpvOpVariable, ptrId, id,
+      std::initializer_list<ir::Operand>{
+          {SPV_OPERAND_TYPE_STORAGE_CLASS, {SpvStorageClassFunction}}}));
+
+  ir::BasicBlock* block = context()->get_instr_block(varInst);
+  block->begin().InsertBefore(std::move(variable));
+  ir::Instruction* inst = &*block->begin();
+
+  // If varInst was initialized, make sure to initialize its replacement.
+  GetOrCreateInitialValue(varInst, index, inst);
+  get_def_use_mgr()->AnalyzeInstDefUse(inst);
+  context()->set_instr_block(inst, block);
+
+  replacements->push_back(inst);
+}
+
+uint32_t ScalarReplacementPass::GetOrCreatePointerType(uint32_t id) {
```

Comment:
This is likely better implemented in analysis::TypeManager.

Reply:
The type manager is pretty bare bones right now. I agree this method would be better implemented elsewhere, but I'd rather have a separate change for that.

Reply:
I take it back, I'll make what changes I can to include the type manager.

Reply:
I'll leave a todo to improve the TypeManager, its not very useful as is.